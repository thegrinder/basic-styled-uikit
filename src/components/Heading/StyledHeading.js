import styled, { css } from 'styled-components';
import { string, bool } from 'prop-types';
import { getBaseLineHeight, calcLineHeight } from '../../theme/typography/typographySelectors';
import {
  getHeadingSizingStyle,
  getHeadingColor,
  getHeadingCommonStyle,
} from './headingSelectors';

const propTypes = {
  color: string.isRequired,
  sizing: string.isRequired,
  marginBottom: bool.isRequired,
};

const StyledHeading = styled.h1`
  display: block;
  ${({
    theme,
    color,
    sizing,
    marginBottom,
  }) => {
    const baseLineHeight = getBaseLineHeight(theme);
    const sizingStyle = getHeadingSizingStyle(theme, sizing);
    return css`
      ${getHeadingCommonStyle(theme, sizing)}
      color: ${getHeadingColor(theme, color)};
      line-height: ${calcLineHeight(sizingStyle.fontSize, baseLineHeight)};
      margin-bottom: ${marginBottom ? `${baseLineHeight}rem` : '0'};
      ${sizingStyle}
    `;
  }}
`;

StyledHeading.propTypes = propTypes;

export default StyledHeading;
