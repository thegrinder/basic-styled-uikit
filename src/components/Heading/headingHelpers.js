import { getStyle } from '../../helpers/utils';

const getHeadingSizing = ({ theme, sizing }) => theme.uiKit.heading.sizings[sizing];
const headingSizingStyles = getStyle(getHeadingSizing);

export default headingSizingStyles;
