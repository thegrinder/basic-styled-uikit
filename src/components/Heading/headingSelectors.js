import { getHeading } from '../../theme/themeSelectors';

export const getHeadingSizingStyle = (theme, sizing) =>
  getHeading(theme).sizings[sizing] || {};

export const getHeadingColor = (theme, color) =>
  getHeading(theme).colors[color];

export const getHeadingCommonStyle = (theme) => getHeading(theme).common;
