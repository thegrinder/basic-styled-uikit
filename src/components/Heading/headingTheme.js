import colors from '../../theme/colors/colorsTheme';

const headingTheme = {
  sizings: {
    h1: {
      fontSize: '3rem',
    },
    h2: {
      fontSize: '2.5rem',
    },
    h3: {
      fontSize: '2rem',
    },
    h4: {
      fontSize: '1.5rem',
    },
    h5: {
      fontSize: '1.25rem',
    },
    h6: {
      fontSize: '1rem',
    },
  },
  colors: {
    white: colors.white,
    light: colors.gray700,
    default: colors.gray800,
    dark: colors.gray900,
    black: colors.black,

    primary: colors.primary500,
    success: colors.success500,
    warning: colors.warning500,
    danger: colors.danger500,
  },
  common: {
    fontWeight: 700,
    fontFamily: 'Lato, Arial, Helvetica, sans-serif',
  },
};

export default headingTheme;
