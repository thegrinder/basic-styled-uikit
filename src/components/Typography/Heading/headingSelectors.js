import { getTypography } from '../typographySelectors';

export const getHeadingSizingStyle = (theme, sizing) =>
  getTypography(theme)?.heading?.sizings?.[sizing];

export const getHeadingColor = (theme, color) => {
  const headingColor = getTypography(theme)?.heading?.colors?.[color];
  const baseColor = getTypography(theme)?.colors?.[color];
  return headingColor || baseColor;
};

export const getHeadingColorOnBg = (theme, bgColor, color) => {
  const typography = getTypography(theme);

  const headingColorOnBg =
    typography?.heading?.onBackground?.[bgColor]?.[color];

  const baseColorOnBg = typography?.onBackground?.[bgColor]?.[color];

  return headingColorOnBg || baseColorOnBg || getHeadingColor(theme, color);
};

export const getHeadingOpacity = (theme, emphasis) =>
  getTypography(theme)?.opacities?.[emphasis];

export const getHeadingCommonStyle = theme =>
  getTypography(theme)?.heading?.common;
