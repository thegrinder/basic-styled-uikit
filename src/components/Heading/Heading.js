import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledHeading from './StyledHeading';

const propTypes = {
  /** determines the size of the heading */
  sizing: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  /** determines the html tag and size, if sizing prop is not specified */
  as: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']).isRequired,
  /** one of: white, light, default, dark, black, primary, success, warning, danger */
  color: string,
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const defaultProps = {
  color: 'default',
  marginBottom: false,
};

const Heading = ({ sizing, color, marginBottom, as, ...rest }) => (
  <StyledHeading
    as={as}
    sizing={sizing || as}
    marginBottom={marginBottom}
    color={color}
    {...rest}
  />
);

Heading.propTypes = propTypes;
Heading.defaultProps = defaultProps;

export default Heading;
