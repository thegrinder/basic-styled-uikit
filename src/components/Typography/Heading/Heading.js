import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledHeading from './StyledHeading';
import { useColorBoxContext } from '../../ColorBox';

const propTypes = {
  /** determines the size of the heading */
  sizing: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  /** one of: high, normal, low */
  emphasis: string,
  /** determines whether colors defined for the background are ignored */
  ignoreBackground: bool,
  /** determines the html tag and size, if sizing prop is not specified */
  as: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']).isRequired,
  /** one of: neutral, primary, success, danger, warning */
  color: string,
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const Heading = ({
  color = 'neutral',
  emphasis = 'high',
  marginBottom = false,
  ignoreBackground = false,
  sizing,
  as,
  ...rest
}) => (
  <StyledHeading
    as={as}
    sizing={sizing || as}
    marginBottom={marginBottom}
    color={color}
    emphasis={emphasis}
    ignoreBackground={ignoreBackground}
    bgColor={useColorBoxContext()}
    {...rest}
  />
);

Heading.propTypes = propTypes;

export default Heading;
