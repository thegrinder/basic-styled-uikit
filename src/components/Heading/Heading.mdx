---
name: Heading
menu: Components
---

import { Playground, Props } from 'docz';
import Heading from './Heading';
import ColorBox from '../ColorBox/ColorBox';

# Heading

<Props of={Heading} />


### Heading sizes

<Playground>
  <Heading as="h1">Heading H1</Heading>
  <Heading as="h2">Heading H2</Heading>
  <Heading as="h3">Heading H3</Heading>
  <Heading as="h4">Heading H4</Heading>
  <Heading as="h5">Heading H5</Heading>
  <Heading as="h6">Heading H6</Heading>
  <Heading as="h6" sizing="h1">Big H6 heading</Heading>
  <Heading as="h1" sizing="h6">Small H1 heading</Heading>
</Playground>


### Heading colors

<Playground>
  <Heading as="h4" color="primary">primary</Heading>
  <Heading as="h4" color="danger">danger</Heading>
  <Heading as="h4" color="success">success</Heading>
  <Heading as="h4" color="warning">warning</Heading>
  <Heading as="h4" color="neutral" emphasis="high">neutral emphasis high</Heading>
  <Heading as="h4" color="neutral" emphasis="normal">neutral emphasis normal</Heading>
  <Heading as="h4" color="neutral" emphasis="low">neutral emphasis low</Heading>
</Playground>

### Heading colors on different backgrounds

<Playground>
  <ColorBox bgColor="primary">
    <Heading as="h4">neutral on primary background</Heading>
  </ColorBox>
  <ColorBox bgColor="success">
    <Heading as="h4">neutral on success background</Heading>
  </ColorBox>
  <ColorBox bgColor="danger">
    <Heading as="h4">neutral on danger background</Heading>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Heading as="h4">neutral on warning background</Heading>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Heading as="h4" ignoreBackground={true}>
      background ignored - default black heading
    </Heading>
  </ColorBox>
</Playground>

### Heading with a bottom margin

<Playground>
  <Heading as="h3" marginBottom={true}>Heading with margin</Heading>
  <p>Sample paragraph.</p>
</Playground>