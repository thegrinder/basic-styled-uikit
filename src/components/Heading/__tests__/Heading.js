import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';

import theme from '../../../theme/theme';
import Heading from '../Heading';

const children = 'children';

const renderComponent = (props = {}) =>
  render(
    <ThemeProvider theme={theme}>
      <Heading as="h1" {...props}>
        {children}
      </Heading>
    </ThemeProvider>
  );

describe('<Heading />', () => {
  it('should render correctly with default props', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toHaveTextContent(children);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render correctly with custom props', () => {
    const {
      container: { firstChild },
    } = renderComponent({
      as: 'h3',
      color: 'primary',
      marginBottom: true,
    });
    expect(firstChild).toBeDefined();
    expect(firstChild).toHaveTextContent(children);
    expect(firstChild).toMatchSnapshot();
  });

  ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].forEach(sizing => {
    it(`should render <${sizing}> tag`, () => {
      const {
        container: { firstChild },
      } = renderComponent({
        as: sizing,
      });
      expect(firstChild.tagName).toEqual(sizing.toUpperCase());
    });
  });

  it('should render should large <h6> heading', () => {
    const {
      container: { firstChild },
    } = renderComponent({
      sizing: 'h1',
      as: 'h6',
    });
    expect(firstChild.tagName).toEqual('H6');
  });
});
