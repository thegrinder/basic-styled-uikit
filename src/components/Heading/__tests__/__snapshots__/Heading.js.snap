// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should render <h1> tag 1`] = `
<h1
  class="StyledHeading-fbinjh-0 fSrojU"
  color="neutral"
>
  children
</h1>
`;

exports[`<Heading /> should render <h2> tag 1`] = `
<h2
  class="StyledHeading-fbinjh-0 csERFs"
  color="neutral"
>
  children
</h2>
`;

exports[`<Heading /> should render <h3> tag 1`] = `
<h3
  class="StyledHeading-fbinjh-0 eznFus"
  color="neutral"
>
  children
</h3>
`;

exports[`<Heading /> should render <h4> tag 1`] = `
<h4
  class="StyledHeading-fbinjh-0 hGqhyU"
  color="neutral"
>
  children
</h4>
`;

exports[`<Heading /> should render <h5> tag 1`] = `
<h5
  class="StyledHeading-fbinjh-0 icpoXR"
  color="neutral"
>
  children
</h5>
`;

exports[`<Heading /> should render <h6> tag 1`] = `
<h6
  class="StyledHeading-fbinjh-0 dYyVrY"
  color="neutral"
>
  children
</h6>
`;

exports[`<Heading /> should render correctly with custom props 1`] = `
<h3
  class="StyledHeading-fbinjh-0 iuCitB"
  color="primary"
>
  children
</h3>
`;

exports[`<Heading /> should render correctly with default props 1`] = `
<h1
  class="StyledHeading-fbinjh-0 fSrojU"
  color="neutral"
>
  children
</h1>
`;

exports[`<Heading /> should render should large <h6> heading 1`] = `
<h6
  class="StyledHeading-fbinjh-0 fSrojU"
  color="neutral"
>
  children
</h6>
`;
